From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: EJ Hsu <ejh@nvidia.com>
Date: Thu, 20 Mar 2025 06:53:20 +0000
Subject: [PATCH 139/243] NVIDIA: SAUCE: xhci: tegra: fix port reset issue

XHCI driver will dynamically power off UTMI pad for root hub ports
entering suspend for power-saving.

If the device triggers a remote wakeup, XHCI driver needs to respond
the resume signal within 1ms according to USB2 spec. This means UTMI
pad power should be powered on within 1ms. However, when USB core driver
receives a "port status change" event for RESUME, it will wait 40ms
to drive the resume signaling on the bus. During this period of time,
XHCI driver has no chance to power on the UTMI pad which leads to the
disconnection of USB device.

In this change, we decide to remove the dynamic powerdown of UTMI pad
for autosuspend.

Bug 4819335
Bug 4752661
Bug 3370448
Bug 5181436

Change-Id: I1704e3b8431c8388c9b5def9ea8a85f073138adf
Signed-off-by: EJ Hsu <ejh@nvidia.com>
Reviewed-on: https://git-master.nvidia.com/r/c/3rdparty/canonical/linux-noble/+/3323167
Reviewed-by: svcacv <svcacv@nvidia.com>
GVS: buildbot_gerritrpt <buildbot_gerritrpt@nvidia.com>
Reviewed-by: WK Tsai <wtsai@nvidia.com>
Reviewed-by: Henry Lin <henryl@nvidia.com>
Reviewed-by: HaoTien Hsu <haotienh@nvidia.com>
(cherry picked from commit 2313d1badb3c649c772874966fef1ed3a3da84f8)

diff --git a/drivers/usb/host/xhci-tegra.c b/drivers/usb/host/xhci-tegra.c
index 6b6bba97385b..5d5e115ec5ef 100644
--- a/drivers/usb/host/xhci-tegra.c
+++ b/drivers/usb/host/xhci-tegra.c
@@ -2915,6 +2915,26 @@ static void tegra_xhci_disable_phy_sleepwalk(struct tegra_xusb *tegra)
 	}
 }
 
+static void tegra_xhci_program_utmi_power_lp0_enter(struct tegra_xusb *tegra)
+{
+	struct xhci_hcd *xhci = hcd_to_xhci(tegra->hcd);
+	struct phy *phy;
+	unsigned int i;
+	u32 portsc;
+
+	for (i = 0; i < xhci->usb2_rhub.num_ports; i++) {
+		if (!xhci->usb2_rhub.ports[i])
+			continue;
+		portsc = xhci_portsc_readl(xhci->usb2_rhub.ports[i]);
+		tegra->lp0_utmi_pad_mask &= ~BIT(i);
+		if (((portsc & PORT_PLS_MASK) == XDEV_U3) || ((portsc & DEV_SPEED_MASK) == XDEV_FS)) {
+			tegra->lp0_utmi_pad_mask |= BIT(i);
+			phy = tegra_xusb_get_phy(tegra, "usb2", i);
+			tegra_phy_xusb_utmi_pad_power_down(phy);
+		}
+	}
+}
+
 static void tegra_xhci_program_utmi_power_lp0_exit(struct tegra_xusb *tegra)
 {
 	unsigned int i, index_to_usb2;
@@ -2945,7 +2965,6 @@ static int tegra_xusb_enter_elpg(struct tegra_xusb *tegra, bool is_auto_resume)
 	unsigned int i;
 	int err;
 	u32 usbcmd;
-	u32 portsc;
 
 	dev_dbg(dev, "entering ELPG\n");
 
@@ -2959,14 +2978,7 @@ static int tegra_xusb_enter_elpg(struct tegra_xusb *tegra, bool is_auto_resume)
 		goto out;
 	}
 
-	for (i = 0; i < xhci->usb2_rhub.num_ports; i++) {
-		if (!xhci->usb2_rhub.ports[i])
-			continue;
-		portsc = xhci_portsc_readl(xhci->usb2_rhub.ports[i]);
-		tegra->lp0_utmi_pad_mask &= ~BIT(i);
-		if (((portsc & PORT_PLS_MASK) == XDEV_U3) || ((portsc & DEV_SPEED_MASK) == XDEV_FS))
-			tegra->lp0_utmi_pad_mask |= BIT(i);
-	}
+	tegra_xhci_program_utmi_power_lp0_enter(tegra);
 
 	err = xhci_suspend(xhci, wakeup);
 	if (err < 0) {
@@ -3063,8 +3075,8 @@ static int tegra_xusb_exit_elpg(struct tegra_xusb *tegra, bool is_auto_resume)
 
 		phy_power_on(tegra->phys[i]);
 	}
-	if (tegra->suspended)
-		tegra_xhci_program_utmi_power_lp0_exit(tegra);
+
+	tegra_xhci_program_utmi_power_lp0_exit(tegra);
 
 	tegra_xusb_config(tegra);
 	tegra_xusb_restore_context(tegra);
@@ -3571,9 +3583,7 @@ static int tegra_xhci_hub_control(struct usb_hcd *hcd, u16 type_req, u16 value,
 	struct tegra_xusb *tegra = dev_get_drvdata(hcd->self.controller);
 	struct xhci_hcd *xhci = hcd_to_xhci(hcd);
 	struct xhci_hub *rhub;
-	struct xhci_bus_state *bus_state;
 	int port = (index & 0xff) - 1;
-	unsigned int i;
 	struct xhci_port **ports;
 	u32 portsc;
 	int ret;
@@ -3587,29 +3597,9 @@ static int tegra_xhci_hub_control(struct usb_hcd *hcd, u16 type_req, u16 value,
 	atomic_inc(&tegra->hub_ctrl_use_cnt);
 
 	rhub = &xhci->usb2_rhub;
-	bus_state = &rhub->bus_state;
-	if (bus_state->resuming_ports && hcd->speed == HCD_USB2) {
-		ports = rhub->ports;
-		i = rhub->num_ports;
-		while (i--) {
-			if (!test_bit(i, &bus_state->resuming_ports))
-				continue;
-			portsc = xhci_portsc_readl(ports[i]);
-			if ((portsc & PORT_PLS_MASK) == XDEV_RESUME)
-				tegra_phy_xusb_utmi_pad_power_on(
-					tegra_xusb_get_phy(tegra, "usb2", (int) i));
-		}
-	}
 
 	if (hcd->speed == HCD_USB2) {
 		phy = tegra_xusb_get_phy(tegra, "usb2", port);
-		if ((type_req == ClearPortFeature) && (value == USB_PORT_FEAT_SUSPEND)) {
-			if (!index || index > rhub->num_ports) {
-				ret = -EPIPE;
-				goto out;
-			}
-			tegra_phy_xusb_utmi_pad_power_on(phy);
-		}
 		if ((type_req == SetPortFeature) && (value == USB_PORT_FEAT_RESET)) {
 			if (!index || index > rhub->num_ports) {
 				ret = -EPIPE;
@@ -3628,11 +3618,6 @@ static int tegra_xhci_hub_control(struct usb_hcd *hcd, u16 type_req, u16 value,
 
 	if (hcd->speed == HCD_USB2) {
 		/* Use phy where we set previously */
-		if ((type_req == SetPortFeature) && (value == USB_PORT_FEAT_SUSPEND))
-			/* We don't suspend the PAD while HNP role swap happens on the OTG port */
-			if (!((hcd->self.otg_port == (port + 1)) && hcd->self.b_hnp_enable))
-				tegra_phy_xusb_utmi_pad_power_down(phy);
-
 		if ((type_req == ClearPortFeature) && (value == USB_PORT_FEAT_C_CONNECTION)) {
 			ports = rhub->ports;
 			portsc = xhci_portsc_readl(ports[port]);
-- 
2.43.0

